docs/typescript.md

```diff
@@ -1,14 +1,209 @@
+
+#### Use `Omit` to reduce requirements
+
+`Omit<T, K>` takes two generic types: `T` representing the original object type and `K` representing the property keys you want to remove. It returns a new type that has all the properties of T except for the ones specified in K. Here are some cases to use omit:
+
+- Removing Unnecessary Properties:
+  Imagine you have a user interface with optional email and phone number fields. However, your API call only cares about the `username`. You can use Omit to reduce the required properties:
+
+```typescript
+interface User {
+  username: string;
+  email?: string;
+  phoneNumber?: string;
+}
+
+// Type for API call payload
+type ApiPayload = Omit<User, 'email' | 'phoneNumber'>;
+
+const payload: ApiPayload = { username: 'johndoe' };
+// Now `payload` only has the `username` property, satisfying the API requirements.
+```
+
+- Conditional Omission:
+  Sometimes, you might want to remove properties based on a condition. `Omit` can still be helpful:
+
+```typescript
+interface CartItem {
+  productId: number;
+  quantity: number;
+  color?: string; // Optional color
+
+// Omit color if quantity is 1
+const singleItemPayload = Omit<CartItem, "color" extends string ? "color" : never>;
+
+// Omit color for all items if quantity is always 1
+const cartPayload: singleItemPayload[] = [];
+```
+
+#### Avoid type assertions, or explain their purpose whenever necessary.
+
+Type assertions inform TypeScript about a variable's actual type, using the `as` syntax.
+While type assertions offer flexibility, they can bypass TypeScript's safety net. This means the compiler won't catch potential type mismatches, leaving you responsible for ensuring correctness.
+
+While it's generally recommended to avoid type assertions in TypeScript, here are two common scenarios where they might be necessary:
+
+1. Using reduce:
+
+Imagine you have an array of User objects and want to find the youngest user using reduce. You start with an empty accumulator `({})`, but the final result should be a User object.
+
+```typescript
+interface User {
+  name: string;
+  age: number;
+}
+
+const users: User[] = [
+  // ... user data
+];
+
+const youngestUser = users.reduce((acc, user) => {
+  if (!acc.age || user.age < acc.age) {
+    acc = user; // Here, `acc` might be undefined initially.
+  }
+  return acc;
+}, {} as User); // Assert `{}` as `User` to satisfy the return type.
+
+console.log(youngestUser); // { name: "...", age: maxAge }
+```
+
+2. Iterating over Enum Keys:
+   When iterating over enum keys using `Object.keys`, the resulting array is of type `string[]`. However, if you want to access enum member values directly, you might need a type assertion.
+
+```typescript
+enum Direction {
+  Up,
+  Down,
+  Left,
+  Right,
+}
+
+const directions: Direction[] = [Direction.Up, Direction.Down];
+
+for (const key of directions) {
+  const direction = directions[key];
+  console.log(direction); // Output: Direction.Up, Direction.Down
+}
+```
```

Comment:
This is no longer an example of `as` assertion being used. Should we replace it with one that fits under the section heading?

> #### Avoid type assertions, or explain their purpose whenever necessary.

Reply:
Ah actually this is my bad. I see this example is working by coincidence: because `Direction.Up` and `Direction.Down` happen to be the first two index keys of the `directions` array.

This would still work as an example of type assertion with the following changes:

```typescript
enum Direction {
  Up = 'up',
  Down = 'down',
  Left = 'left',
  Right = 'right',
}

const directions: Direction[] = [Direction.Up, Direction.Down];

for (const key of directions) {
  const direction = directions[key as keyof typeof directions];
  console.log(direction); // Output: Direction.Up, Direction.Down
}
```

We also have a built-in utility type for narrowing key signatures, which simplifies the code.

```ts
import { getKnownPropertyNames } from '@metamask/utils';

...

for (const key of getKnonwPropertyNames(direction)) {
  const direction = directions[key];
}
```

Reply:
Thanks for the suggestion, i've learnt a new`getKnonwPropertyNames` today